docs: refactor Spanner DB guide (#65118)
gravitational/teleport: docs/pages/enroll-resources/database-access/enrollment/google-cloud/spanner.mdx

The **Cloud Spanner Database User** role is predefined. You can create custom IAM roles to further restrict access if needed.
</Admonition>

### Allow service account impersonation

Grant the Database Service account permission to impersonate the database user account.

<Tabs>
<TabItem label="Google Cloud console">

1. Navigate to **IAM & Admin** → **Service Accounts**
1. Select `spanner-user`
1. Open the **Permissions** tab
1. Click **Grant Access**
1. Add the `teleport-db-service` account
1. Assign **Service Account Token Creator**
1. Click **Save**

</TabItem>

<TabItem label="gcloud CLI">

```code
$ gcloud iam service-accounts add-iam-policy-binding \
  spanner-user@<Var name="project-id" />.iam.gserviceaccount.com \
  --member="serviceAccount:teleport-db-service@<Var name="project-id" />.iam.gserviceaccount.com" \
  --role="roles/iam.serviceAccountTokenCreator"
```

</TabItem>
</Tabs>

<Checkpoint
  title="Verify GCP permissions"
  description="Confirm that the Database Service account can impersonate the database user account."
>

Run:

```code
$ gcloud iam service-accounts get-iam-policy \
  spanner-user@<Var name="project-id" />.iam.gserviceaccount.com \

>   --format=json | grep "roles/iam.serviceAccountTokenCreator"

Author: updated the query to make it more narrow

Provide the following information and then generate a configuration file for the
Teleport Database Service:

- <Var name="example.teleport.sh:443" /> The host **and port** of your Teleport Proxy Service
or cloud-hosted Teleport Enterprise site.
- <Var name="project-id"/> The GCP project ID. You can normally see it in the
organization view at the top of the GCP dashboard.
- <Var name="instance-id"/> The name of your Cloud Spanner instance.

```code
$ sudo teleport db configure create \
  -o file \
  --name=spanner-example \
  --protocol=spanner \
  --labels=env=dev \
  --token=/tmp/token \
  --uri=spanner.googleapis.com:443 \
  --proxy=<Var name="example.teleport.sh:443" /> \
  --gcp-project-id=<Var name="project-id" /> \
  --gcp-instance-id=<Var name="instance-id" />
```

## Step 4/6. Configure GCP credentials

(!docs/pages/includes/database-access/cloudsql_service_credentials.mdx serviceAccount="teleport-db-service"!)

## Step 5/6. Start the Teleport Database Service

(!docs/pages/includes/start-teleport.mdx service="the Teleport Database Service"!)

<Checkpoint
  title="Verify database availability"

>   description="Confirm that the Spanner database appears in Teleport."

Reviewer: I would add a more specific description of what the user should do before they can verify this, e.g., instructions to visit the Resources page on the Web UI or run `tsh db ls` command.

  --labels=env=dev \
  --token=/tmp/token \
  --uri=spanner.googleapis.com:443 \
  --proxy=<Var name="example.teleport.sh:443" /> \
  --gcp-project-id=<Var name="project-id" /> \
  --gcp-instance-id=<Var name="instance-id" />
```

## Step 4/6: Configure GCP credentials

(!docs/pages/includes/database-access/cloudsql_service_credentials.mdx serviceAccount="teleport-db-service"!)

## Step 5/6: Start the Teleport Database Service

(!docs/pages/includes/start-teleport.mdx service="the Teleport Database Service"!)

<Checkpoint
  title="Verify database availability"
  description="Confirm that the Spanner database appears in Teleport."
>

Replace `example.teleport.sh` and `example-user` with the proxy and username in question.

```code
$ tsh login --proxy=<Var name="example.teleport.sh" /> --user=<Var name="example-user" />
$ tsh db ls
```

You should see:

```
Name            Description       Allowed Users Labels  Connect
--------------- ----------------- ------------- ------- -------
spanner-example GCP Cloud Spanner [*]           env=dev
```

If the database does not appear, verify your Teleport role permissions. See the [RBAC](../../rbac.mdx) guide.

</Checkpoint>

> ## Step 6/6: Connect using the service account

Author: Added a small paragraph:

In this step, you will create a Teleport user with access to the Spanner
database, connect using the impersonated service account, and optionally
remove stored credentials when you are done.

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- A Google Cloud account
- A Cloud Spanner instance and database
- A host to run the Teleport Database Service (for example, a Compute Engine instance) 
  with outbound HTTPS access to `spanner.googleapis.com:443`
- (!docs/pages/includes/tctl.mdx!)

## Step 1/6. Create service accounts

Teleport requires two service accounts:

- **Database Service account**: used by the Teleport Database Service to access Google Cloud
- **Database user account**: impersonated per connection to represent the database identity

> For example, in the following steps, we will create `teleport-db-service` and `spanner-user` respectively.

Reviewer 2: There's some ambiguity here because "service account" is a thing in GCP, and "database service" is referring to a Teleport db agent.

Also, there is only one service account for the agent, but can be many service accounts representing different database users.

nit: "respectively" doesn't feel right in the last sentence IMO

```suggestion
Teleport requires two types of Google Cloud service accounts:

- **Database Service**: a single service account used by the Teleport Database Service to access Google Cloud and impersonate database user service accounts
- **Database user**: one or more service accounts impersonated per connection to represent database users

For example, in the following steps, we create `teleport-db-service` (Database Service service account) and `spanner-user` (a Database user service account).
```

### Create the Database Service account

This account is used by the Teleport Database Service to access Google Cloud and impersonate the database user account.

<Tabs>
<TabItem label="Google Cloud console">

1. Navigate to **IAM & Admin** → **Service Accounts**.
1. Click **Create Service Account**.
1. Set the name to `teleport-db-service`.
1. Skip optional steps and click **Done**. You will grant permissions later.

</TabItem>

<TabItem label="gcloud CLI">

```code
$ gcloud iam service-accounts create teleport-db-service --display-name="Teleport Database Service"
```

</TabItem>
</Tabs>

> ### Create the database identity account

Reviewer 2: ```suggestion
### Create the database user account
```

> This account represents the database identity. Teleport users specify this name when connecting.

Reviewer 2: ```suggestion
This account represents a database user. Teleport users specify this name when connecting.
```

  --gcp-project-id=<Var name="project-id" /> \
  --gcp-instance-id=<Var name="instance-id" />
```

## Step 4/6. Configure GCP credentials

(!docs/pages/includes/database-access/cloudsql_service_credentials.mdx serviceAccount="teleport-db-service"!)

## Step 5/6. Start the Teleport Database Service

(!docs/pages/includes/start-teleport.mdx service="the Teleport Database Service"!)

<Checkpoint
  title="Verify database availability"
  description="Confirm that the Spanner database appears in Teleport."
>

Replace <Var name="example.teleport.sh" /> and <Var name="example-user" /> with the proxy and username in question.

```code
$ tsh login --proxy=<Var name="example.teleport.sh" /> --user=<Var name="example-user" />
$ tsh db ls
```

You should see:

```
Name            Description       Allowed Users Labels  Connect
--------------- ----------------- ------------- ------- -------
spanner-example GCP Cloud Spanner [*]           env=dev
```

If the database does not appear, verify your Teleport role permissions. See the [RBAC](../../rbac.mdx) guide.

</Checkpoint>

## Step 6/6. Connect using the service account

In this step, you will create a Teleport user with access to the Spanner

> database, connect using the impersonated service account, and optionally

Reviewer 2: ```suggestion
database, connect using the database user service account, and optionally
```

remove stored credentials when you are done.

(!docs/pages/includes/database-access/create-user.mdx!)

Log in to your Teleport cluster as the new user that you just created:
```code
$ tsh db connect --db-user=spanner-user --db-name=example-db spanner-example
$ tsh login --proxy=<Var name="example.teleport.sh" /> --user=<Var name="example-user" />
```

(!docs/pages/includes/database-access/proxy-db-tunnel.mdx db="spanner-example" dbArgs="--db-user=spanner-user --db-name=example-db"!)
Connect using the service account name (without the domain suffix):
(Note: `--db-name` here refers to the specific database ID within your Spanner instance.)

To log out of the database and remove credentials:
```code
$ tsh db connect --db-user=spanner-user --db-name=database-name spanner-example
```

You should now have an authenticated connection using the impersonated service account.

To remove credentials:

```code
# Remove credentials for a particular database instance:
$ tsh db logout spanner-example
# Or remove credentials for all databases:
```

To remove all database credentials:

```code
$ tsh db logout
```

## Troubleshooting

> (!docs/pages/includes/database-access/gcp-troubleshooting.mdx!)

Reviewer 2: Why are we inlining this include?

Author: Reverting this; I think the include could use some slight edits but I'll save it for another PR.
